source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,792 @@
+
+    task->timestamp = run_at_nanos;
+
+    /* Begin critical section */
+    aws_mutex_lock(&impl->cross_thread_data.mutex);
+
+    bool should_signal_thread = aws_linked_list_empty(&impl->cross_thread_data.tasks_to_schedule);
+    aws_linked_list_push_back(&impl->cross_thread_data.tasks_to_schedule, &task->node);
+
+    /* Signal thread that cross_thread_data has changed (unless it's been signaled already) */
+    if (should_signal_thread) {
+        uv_async_send(&impl->schedule_tasks_async);
+    }
+
+    aws_mutex_unlock(&impl->cross_thread_data.mutex);
+    /* End critical section */
+}
+
+static void s_schedule_task_now(struct aws_event_loop *event_loop, struct aws_task *task) {
+    s_schedule_task_common(event_loop, task, 0 /* zero denotes "now" task */);
+}
+
+static void s_schedule_task_future(struct aws_event_loop *event_loop, struct aws_task *task, uint64_t run_at_nanos) {
+    s_schedule_task_common(event_loop, task, run_at_nanos);
+}
+
+static void s_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task) {
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    struct aws_hash_element *elem = NULL;
+    aws_hash_table_find(&impl->el_thread_data.running_tasks, task, &elem);
+    if (elem) {
+        struct task_data *task_data = elem->value;
+
+        /* Remove handle from hash table to prevent future collsions when rescheduling */
+        int was_present = 0;
+        aws_hash_table_remove(&impl->el_thread_data.running_tasks, task_data->task, NULL, &was_present);
+        assert(was_present);
+
+        aws_task_run(task_data->task, AWS_TASK_STATUS_CANCELED);
+
+        uv_timer_stop(&task_data->timer);
+        uv_close((uv_handle_t *)&task_data->timer, s_uv_close_timer);
+    }
+}
+
+static void s_uv_close_sub_async(uv_handle_t *handle) {
+
+    struct handle_data *handle_data = handle->data;
+    struct libuv_loop *impl = handle_data->event_loop->impl_data;
+
+    handle->data = impl;
+    s_uv_close_handle(handle);
+}
+
+static void s_uv_async_poll_start(uv_async_t *handle) {
+
+    struct handle_data *handle_data = handle->data;
+    struct libuv_loop *impl = handle_data->event_loop->impl_data;
+
+    uv_poll_init(impl->uv_loop, &handle_data->poll, handle_data->owner->data.fd);
+    uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb);
+
+    uv_close((uv_handle_t *)handle, s_uv_close_sub_async);
+}
+
+static int s_subscribe_to_io_events(
+    struct aws_event_loop *event_loop,
+    struct aws_io_handle *handle,
+    int events,
+    aws_event_loop_on_event_fn *on_event,
+    void *user_data) {
+
+    assert(event_loop);
+    assert(handle->data.fd != -1);
+    assert(handle->additional_data == NULL);
+    assert(on_event);
+    /* Must subscribe for read, write, or both */
+    assert(events & (AWS_IO_EVENT_TYPE_READABLE | AWS_IO_EVENT_TYPE_WRITABLE));
+
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
```

Comment:
Do we want to do less in the case that the event loop isn't currently running?
ex: don't increment num_open_handles, don't register with uv, etc...